distribution/src/config/elasticsearch-plugins.example.yml

```diff
@@ -0,0 +1,26 @@
+# Rename this file to `elasticsearch-plugins.yml` to use it.
+#
+# All plugins must be listed here. If you add a plugin to this list and run
+# `elasticsearch-plugin sync`, that plugin will be installed. If you remove
+# a plugin and re-run the command, that plugin will be removed.
+
+plugins:
+  # Each plugin must have an ID. Plugins with only an ID are official plugins and will be downloaded from Elastic.
+  - id: example-id
+
+  # Plugins can be specified by URL (it doesn't have to be HTTP, you could use e.g. `file:`)
+  - id: example-with-url
+    url: https://some.domain/path/example4.zip
+
+  # Or by maven coordinates:
+  - id: example-with-maven-url
+    url: org.elasticsearch.plugins:example-plugin:1.2.3
```

Comment:
Or rather, is there a reason why we need both? Currently when you use the plugin CLI users simply provide either a) and id for "official" plugins b) a uri for third party plugins or c) maven coordinates. The the latter cases, what's the purpose of having them also specify an ID, especially since we don't require that when using the normal `install` command. Should we simplify and just ask for a collection of "descriptors" which can be any of the above?

Reply:
The issue is that the sync command can't know for sure what a plugin's name / ID is until it's downloaded it, so it needs that information up-front so that it can determine whether the plugin is already installed. Otherwise it would be trying to reconcile a list of locations with a list of directory names from the `plugins/` directory. We don't have that issue with the install or remove commands because they are imperative actions.

Note that I don't track whether a plugin's URL has changed, and so don't try to "upgrade" plugins. I figured we'd add that if we ever actually needed it.

Reply:
Ah, that makes sense now.

> Note that I don't track whether a plugin's URL has changed, and so don't try to "upgrade" plugins. I figured we'd add that if we ever actually needed it.

So how _do_ users upgrade plugins? As of right now do they just run the `install` command again?

Reply:
Remove and install, according to the docs. We'd need some way of detecting a change in locations between executions of the tool, and then run a remove and add. I think I will actually need to add something here, because otherwise you'd need to edit the config, sync, edit it and sync again, and that's rubbish (and probably wouldn't work in Cloud).

Reply:
Could we add a dot file in each installed plugin dir showing the source of how it was installed? Then you could know for listing the plugins dirs where each came from. Alternatively, we could have a single metadata file at the root of the plugins dir.

Reply:
If it’s a single file, you could also diff the last changed time with the configuration file last changed to know when action is needed.

Reply:
I think making this part of bootstrap make some of this moot. We'll simply check on each startup if we need to do something. For official plugins, we'll upgrade them, for non-official ones that are out of date we can log a warning.